feat(apollo): Create removeExercise resolver

flacial opened the pull request:
## Motivation
Currently, there is no way to set the removed flag on an exercise when an admin wants to remove it. The removed flag is a field in the database that indicates whether an exercise has been removed or not.

The PR is adding a resolver that will allow an admin to set the removed flag to true when they want to remove an exercise. This will mark the exercise as removed in the database, and it will no longer be visible to users.

## Steps taken
1. Created the resolver `removeExercise`
2. Ensure everything is working as intended by adding tests for the resolver
3. Created the query file for `removeExercise` resolver so it can be easily used in the client-side
4. Update GraphQL `index.ts` with the new resolver and hooks.

### Unrelated steps
- Fixes some typos

## Resolver logic
1. It receives one argument, the user `id`.
2. It fetches the exercise to be removed from the database using the provided `id`.
3. It retrieves the `id` of the user making the request (`authorId`) from the request context (req).
4. If the user is not an admin and the `authorId` does not match the `authorId` of the exercise being removed, the resolver throws an error, indicating that the user is not authorized to remove the exercise.
6. If the exercise is already removed, the resolver throws an error, indicating that exercise is already flagged.
7. If the user is authorized to remove the exercise, the resolver updates the exercise in the database, setting the `removedAt` and `removedById` fields to new Date & the user ID. It also includes the author and module relations in the update.
8. The resolver returns the updated exercise.

## Testing results

Executes `removeExercise`:
```gql
mutation {
  removeExercise(id: 1) {
    id
  }
}
```

**Result:**
```gql
{
  "data": {
    "removeExercise": {
      "id": 1,
      "removedAt": "1672591865633",
      "removedById": 1
    }
  }
}
```

**Result if it's already removed:**
```gql
{
  "errors": [
    {
      "message": "Exercise is already removed",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
  "data": null
}
```

**Result if the exercise is not found:**
```gql
{
  "errors": [
    {
      "message": "Exercise is not found",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ]
  ],
  "data": null
}
```

## Testing instructions
1. Go to `/api/graphql`
2. Run the queries from `Testing result` section
3. Make sure they return the correct results

## Related issues
- #2630

JasirZaeem commented on graphql/resolvers/exerciseCrud.ts:
```diff
@@ -102,6 +104,41 @@ export const deleteExercise = withUserContainer<
+
+  const authorId = req.user?.id
+
+  if (!isAdmin(req) && exercise?.authorId !== authorId)
```
This if has no curly braces, the next one does, both have single line statements. This is just a style suggestion but can you make them uniform (preferably add braces to both)

JasirZaeem commented:
This is out of the scope of this particular PR but related, maybe removed should be removedAt/removed_at as well like some other booleans we have changed to date | null before.
Like flagged at and the user who flagged it, it would make sense for removed to be like this too with removed at and removed by data.
And even if removed remains a boolean it should not be an optional field, it should either be required with false as default so only two values are possible, or it should be a date field with default value being date. Currently removed can hold three values null, false, true.

flacial commented:
> This is out of the scope of this particular PR but related, maybe removed should be removedAt/removed_at as well like some other booleans we have changed to date | null before. Like flagged at and the user who flagged it, it would make sense for removed to be like this too with removed at and removed by data. And even if removed remains a boolean it should not be an optional field, it should either be required with false as default so only two values are possible, or it should be a date field with default value being date. Currently removed can hold three values null, false, true. My suggestion would be to refactor the schema for exercise and improve it with regards to remove before this PR.

I agree. It makes sense.

JasirZaeem commented on graphql/resolvers/exerciseCrud.ts:
```diff
@@ -102,6 +103,46 @@ export const deleteExercise = withUserContainer<
+  return prisma.exercise.update({
+    where: { id },
+    data: {
+      removedAt: new Date().toISOString(),
```
You don't need to convert to isostring, just passing a javascript date object is enough prisma will serialise it internally
e.g.
https://github.com/garageScript/c0d3-app/blob/cd8d8a595098ff376f734328fd08623d3c7b12a3/helpers/updateSubmission.ts#L108-L119
